x-pack/plugins/security/server/lib/authorization/has_privileges.test.js

```diff
@@ -43,224 +37,507 @@ const createMockServer = ({ settings = {} } = {}) => {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(callWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      index: [{
+        names: [ defaultKibanaIndex ],
+        privileges: ['read', 'index']
+      }]
+    }
+  });
+  expect(result).toEqual({
+    success: false,
+    missing: [getLoginPrivilege(), ...privileges],
+    username,
+  });
 });
 
-test(`excludes granted privileges from missing privileges`, async () => {
+//eslint-disable-next-line max-len
+test(`returns success of true if the user doesn't have any application privileges but they have index privilege on kibana index`, async () => {
+  const privilege = 'something-completely-arbitrary';
+  const username = 'foo-username';
   const mockServer = createMockServer();
-  mockResponse(false, {
-    [getVersionPrivilege(defaultVersion)]: true,
-    [getLoginPrivilege()]: true,
-    foo: false,
-    bar: true,
-  });
+  const callWithRequest = createMockCallWithRequest([
+    mockApplicationPrivilegeResponse({
+      hasAllRequested: false,
+      privileges: {
+        [getVersionPrivilege(defaultVersion)]: false,
+        [getLoginPrivilege()]: false,
+        [privilege]: false,
+      },
+      username,
+    }),
+    mockLegacyResponse({
+      hasAllRequested: false,
+      privileges: {
+        read: false,
+        index: true,
+      },
+      username,
+    })
+  ]);
 
   const hasPrivilegesWithRequest = hasPrivilegesWithServer(mockServer);
-  const hasPrivileges = hasPrivilegesWithRequest({});
-  const result = await hasPrivileges(['foo']);
-  expect(result.missing).toEqual(['foo']);
+  const request = Symbol();
+  const hasPrivileges = hasPrivilegesWithRequest(request);
+  const privileges = ['foo'];
+  const result = await hasPrivileges(privileges);
+
+  expect(callWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(callWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      index: [{
+        names: [ defaultKibanaIndex ],
+        privileges: ['read', 'index']
+      }]
+    }
+  });
+  expect(result).toEqual({
+    success: true,
+    missing: [],
+    username,
+  });
 });
 
-test(`throws error if missing version privilege and has login privilege`, async () => {
+//eslint-disable-next-line max-len
+test(`returns success of false if the user doesn't have any application privileges and they have the read privilege on kibana but the privilege isn't a read action`, async () => {
```

Comment:
nit: would these test names be better understood if they were nested in `describe` blocks? It could cut down on the verbosity here, while still giving an understandable output in the test report

Reply:
Probably :) I'll play around with this a bit more